### ISSUE TITLE
RFC: Query/Mutation/Subscription Pre-Processor

### ISSUE THREAD

**Issue by nyteshade** [apollographql/apollo-client#10228]
## Target
JavaScript / TypeScript Apollo Client
_other platforms would ideally support as well_

## Description
I have the pleasure of developing a React Native mobile application that has both offline and online needs regarding its data dependencies. Apollo supports this usually with the [`@client` directive](https://www.apollographql.com/docs/react/development-testing/client-schema-mocking/#3-query-with-the-client-directive) and as long as the type or field does not reside on the remote server this isn't a problem.

The problem comes into play when there are needs for the data to exist both remotely and locally. To this end I have put together a proof of concept that allows me to solve this problem, but without changing every query that exists out there when the application is offline vs. online, the query, mutation or subscription needs to be modified before it is executed.

The proof of concept introduces a local `@persistent` directive that is not actually registered in either the remote or local `typeDefs` SDL. Instead queries would be modified one time for fields that have a local backing (in SQLite for us). This new `@persistent` directive gets swapped for a `@client` directive when the application is offline and removed entirely when the application is online. 

This swapping of directives occurs in the proposed lifecycle of an operation. That proposal is to have the ability to register a pre-op processor. Currently, this is achieved in the proof of concept by wrapping the `.query`, `.mutate` and `.subscribe` methods of the Apollo Client with higher order functions that invoke the originals after processing the SDL first. 

Here is an example of the `.mutate` pre-processor. It is almost identical for `.query` and `.subscribe`

```js
client[Symbol.for('mutate')] = client.mutate

client.mutate = function (options) {
  // create a copy of the supplied SDL/AST to prevent mutation
  let newOpSDL = gql`${print(options.mutation)}`

  newOpSDL = swapDirective(
    newOpSDL,
    'persistent',
    () => online() ? undefined : 'client'
  ) || options.mutation

  return client[Symbol.for('mutate')]({ ...options, mutation: newOpSDL })
}
```

The proof of concept `swapDirective()` function is here. The optional chaining technique would make this code cleaner but this is a proof of concept and it wasn't written that way.

```js
/****************************************************************************
 * Create a function that will swap out a directive with a new value in
 * a given GQL / SDL document object.
 *
 * @param {GQL.ASTNode} document the document to search for a directive in
 * @param {String|Function} searchFor a String or function that returns a
 * string, that indicates the directive to search for
 * @param {String|Function} replaceWith a String or function that returns a
 * string, that indicates the directive value to swap in. Note that if this
 * value is not-truthy the directive will be removed.
 * @return an altered document if changes were made or null otherwise
 ****************************************************************************/
 function swapDirective(document, searchFor, replaceWith) {
  let alteredDocument = document

  const isFn = (o) => /Function/.test(Object.prototype.toString.call(o))
  const target = isFn(searchFor) ? searchFor() : String(searchFor)
  const newValue = isFn(replaceWith) ? replaceWith() : String(replaceWith)

  const getter = (n) => n && n.name && n.name.value || null
  const setter = (n, v) => { n && n.name && (n.name.value = v); return n }
  const hasTarget = (n) => new RegExp(target, 'i').test(getter(n))

  alteredDocument = visit(document, {
    Field: {
      enter(node, key, parent, path, ancestors) {
        if (node.directives && node.directives.some(hasTarget)) {
          if (!!!newValue) {
            for (let i = 0; i < node.directives.length; i++) {
              if (new RegExp(target, 'i').test(getter(node.directives[i]))) {
                node.directives.splice(i, 1)
                i--
              }
            }
          }
          else {
            // Thought process here is to swap @persistent for @client
            // when we are offline.
            node.directives = node.directives
              .filter(directive => new RegExp(target, 'i').test(getter(directive)))
              .map(directive => {
                return setter(directive, newValue)
              })
          }
          return node
        }
      },
    }
  })

  return alteredDocument || document
}
```

## Proposal
The idea here would be to have the Apollo Client allow both the registration and de-registration of an AST/SDL preprocessor. It would, ideally, also allow reorganization of the order in which the preprocessor could be arranged. Registered function would process the SDL/ASTs provided before they are used normally by the Apollo GraphQL Client code.

It may work and function as seen here
```js
let client = new ApolloClient(...);

// Grab some constant values from the client class (should maybe just also have a PROCESS_ALL constant)
let { PROCESS_QUERY, PROCESS_MUTATION, PROCESS_SUBSCRIPTION } = ApolloClient;

// Choose which items our preprocessor applies to
let processedOperations = PROCESS_QUERY | PROCESS_MUTATION | PROCESS_SUBSCRIPTION;

// Create a function to do the pre-processing
function processor(sdl, ast) {
  // create a copy of the supplied SDL/AST to prevent mutation
  let newOpSDL = gql`${print(ast)}`

  newOpSDL = swapDirective(
    newOpSDL,
    'persistent',
    () => online() ? undefined : 'client'
  ) || ast

  return client[Symbol.for('mutate')]({ ...options, mutation: newOpSDL })
}

client.registerPreprocessor(processedOperations, processor);

// Perform queries, mutations or subscriptions normally
await client.query(...);
await client.mutate(...);
await client.subscribe(...);

// Optionally remove the preprocessor for all calls to `.mutate`
client.deregisterPreprocessor(PROCESS_MUTATION, processor);

// etc...
```

**Comment by jpvajda** [comment]
The issue is now moved thanks to @hwillson 🙏

**Comment by nyteshade** [comment]
Understood, I'll try to play with ApolloLink in a vacuum a bit more to understand its pros and cons and how it might be used by the client in this case. Please let me know if there are relevant places for me to peek at @jpvajda @bignimbus. The ApolloLink documentation linked above and found by searching shows how a consumer consumes and works with the links. It is not as enlightening when describing how the links are used and processed internally by the client. 

It seems they use the Observer/Observable system from zen-observable. It is not clear how compatible or incompatible this is with async/await promise based code or what exactly to do with ApolloLink in the format of an end link given that it won't be making any HTTP requests. I have some ideas, but this is where I am in the ideation process so far.

Thank you @hwillson for moving the issue to the correct location

**Comment by nyteshade** [comment]
Updates:

I have contributed some changes to the `zen-observable` repo used by the Apollo client. I've messaged @benjamn on the `zen-observable-ts` repo with some questions on how to get a PR merged there. I'm working on the changes needed for the client using ApolloLink but its still a WIP. 

Writing production ready code takes more time than proof-of-concept code. So I'm trying to figure out the best insertion points to have an ApolloLink preprocessor code path be introduced. I think I have it but I'm not done. Will keep folks informed.

**Comment by hwillson** [comment]
Thanks for working on all of this @nyteshade! We'll take a closer look at https://github.com/apollographql/zen-observable-ts/issues/310 and get back to you shortly (Ben is OOO 🍹).

Just a quick note regarding:

> ApolloLink would be an ideal solution, but it seems to have already stripped out all of the `@client` portions of the schema before there is any opportunity to process the document in my case.

We intentionally moved away from `@client` being available in the link chain when we moved local resolvers into the core. We've since come up with a newer API that can be used with `@client` (type / field policies), and are recommending people use this newer API when they can. At some point in the future we'll likely fully remove local resolvers, but if we do that we want to make sure folks can use `@client` in the link chain, if they want to re-create local resolver behaviour for example. All of this being said, we're tracking the re-adding of `@client` into the link chain [here](https://github.com/apollographql/apollo-client/issues/10060) which would help your use case for sure.

**Comment by nyteshade** [comment]
So I have a functional PR that I am just working on adjusting the unit tests for. Shall I put it up for review or not? If this is likely not wanted due to a different direction you're working from internally, we'll have to monkey patch the code for our own needs in the mean time which would be less desirable (even if it were faster). Also Zach Steele and several others reassured at the GraphQL Summit that local client resolvers and code would not be going away any longer. Have I missed something @hwillson?

**Comment by jpvajda** [comment]
@nyteshade Thank you! We'll chat further tomorrow with you on this. Looking forward to the discussion. 🙏

**Comment by jerelmiller** [comment]
Hey @nyteshade! As per our discussion earlier, we'd like to introduce a new boolean option that enables `@client` directives to make it to the link chain so that you can do the processing you need on the query. I've opened a new issue (https://github.com/apollographql/apollo-client/issues/10303) to track this as we see it as part of the work to move local resolvers back to the link chain (#10060).

As such, I'm going to close this RFC. We sincerely appreciate your thoughts and discussion as its really helped us figure out how to move forward with this. @jpvajda will follow up with you to set expectations on timing and how we plan to release this change. If you have followup questions, feel free to post them in #10303!

**Comment by jerelmiller** [comment]
Hey @nyteshade! I just opened https://github.com/apollographql/apollo-client/pull/10346 which adds the ability to get `@client` fields in the link chain. Once that PR is merged, we will release it in an upcoming `3.8.0-alpha.x` version.

Feel free to take a look at the PR and add any comments/questions/feedback. I hope this will work well for your use case!

**Comment by adamesque** [comment]
Hey @jerelmiller — I'm interested in the original goal of the RFC, which was targeted at adding first-class support for operation manipulation in a Link. I've [mentioned it elsewhere](https://github.com/apollographql/apollo-client/pull/9764#issuecomment-1306150181), but we're planning on building at least one new client-only directive to support referencing remote query fragments via webpack module federation.

It looks like y'all took a much more targeted approach (optionally exposing `@client` directives to Links), but I'm curious if there's any appetite for the broader feature. I see in #10346 you added a `transformQuery` option to the client constructor, which could be a natural place to register transform functions (alongside specific named transforms). There's also another [open feature request](https://github.com/apollographql/apollo-feature-requests/issues/166) in addition to this RFC requesting similar functionality.

Anyway, I love that you solved for this specific use-case, but don't want to lose track of the broader request either!

**Comment by jerelmiller** [comment]
Hey @adamesque!

That's a really great question! As I understand the original proposal, the `@persistent` directive mentioned in the description was a way for an Apollo Link to determine whether the app was online/offline. In the case the app was online, the link would remove the `@persistent` directive and the query would be requested from the server as normal. If the app was offline, it would execute `@client` queries to query a local data store to pull data from it to display to a user.

As we talked about this particular feature, we determined that `@client` being stripped in Apollo core before making it to the link chain is one the reasons the above use case was difficult. Apollo Link is the natural network extension point for Apollo Client, so it makes sense this kind of manipulation could be done in a link. 

This is part of a broader effort to [bring back local resolvers](https://github.com/apollographql/apollo-client/issues/10060), which were deprecated at some point during the 3.x series. The more we talked about it, the more it seemed there was no reason we need to deprecate this API entirely. 

To be transparent, the `transformQuery.removeClientFields` option is more-or-less a stopgap solution until we can make breaking changes in Apollo v4, where we'd be able to move all of this logic to a custom Apollo link. The idea was that once v4 was out, we'd be able to remove this newly created option and rely on the behavior in the link.

That's not to say we can't use `transformQuery` options for anything else. You make a really good point that this could be a good place to add custom query transforms. I think we may still need to think through that a bit more to determine how to attach custom logic to the transform in a natural way. I think of things like being able to query the cache, get access to query results, etc are things that this API would need to be able to handle. Unfortunately the PR I added is a bit too simplistic as it stands.

I'll think through this a bit more and talk with the Apollo Client team to determine the best course of action. Since that logic is currently released in an alpha, there is a chance we could make this type of change between now and the official 3.8 release. We will definitely think through it a bit more and determine the path we want to go. At the very least, just be aware these changes were made in the broader scope of being able to add local resolvers in the link chain.

**Comment by jerelmiller** [comment]
@adamesque I just opened a new proposal (https://github.com/apollographql/apollo-client/issues/10481) for a broader query transforms API. I think this might address some of what you're looking for. I'm not sure it addresses all of it, but it at least might be a first step toward that goal. Take a look when you get a chance and let me know if you have any feedback. Thanks!

**Comment by adamesque** [comment]
@jerelmiller oh wow! I am VERY excited to read this! Thank you, and I'll try to get you feedback next week. Much appreciated!
